### PR TITLE
Code documentation in the form of Javadocs and comments

### PR DESCRIPTION
**Pull request by bertfrees**
This contains all the changes I did in https://github.com/brailleapps/dotify.formatter.impl/pull/102, with some improvements, and one or two changes from Paul are also included.

The improvements compared with https://github.com/brailleapps/dotify.formatter.impl/pull/102 are:

- I changed all the occurences of "used by X" in class Y by "uses Y" in class X. Joel made a remark about that last time.
- I improved the overview page to make it easier to navigate to certain parts of the code.
- I fixed my remaining TODOs.
- I answered some more of Paul's questions.
- Some other additions and rewordings.

@kalaspuffar, @PaulRambags and @joeha480 What do you think? Is this OK to merge? We're still a long way from a full coverage, and things can be expanded and worded differently, but I think this is already a major improvement. We can further improve the documentation of specific parts of the code as we work on that code. 

The remaining questions from Paul in https://github.com/brailleapps/dotify.formatter.impl/pull/102 (which is not much anymore) can be discussed and handled in a subsequent PR.

### CODE REVIEW

File: src/org/daisy/dotify/formatter/impl/sheet/EvenSizeVolumeSplitter.java [hunk 1]
```diff
@@ -4,10 +4,31 @@
...
+ * </ul>
+ *
+ * <p>The actual sizes of the volumes are not determined here. This is done in {@link
+ * org.daisy.dotify.formatter.impl.VolumeProvider}.</p>
+ */
 class EvenSizeVolumeSplitter implements VolumeSplitter {
 	private static final Logger logger = Logger.getLogger(EvenSizeVolumeSplitter.class.getCanonicalName());
 	private EvenSizeVolumeSplitterCalculator sdc;
 	private final SplitterLimit splitterMax;
+	// number of volumes to add on top of the number of volumes strictly required to contain the
```
Comment (kalaspuffar) [inline src/org/daisy/dotify/formatter/impl/sheet/EvenSizeVolumeSplitter.java]: I tried to understand this sentence and it can be because of lack of context but I don't understand.

File: src/org/daisy/dotify/formatter/impl/page/PageSequenceBuilder2.java [hunk 2]
```diff
@@ -60,14 +92,40 @@
...
-	public PageSequenceBuilder2(int fromIndex, LayoutMaster master, int pageOffset, BlockSequence seq, FormatterContext context, DefaultContext rcontext, SequenceId seqId, BlockLineLocation blc) {
+	/**
+	 * @param fromIndex "Index" of the first page of this <code>PageSequenceBuilder</code>. The
+	 *     exact value does not matter for <code>PageSequenceBuilder2</code> itself, but {@link
+	 *     org.daisy.dotify.formatter.impl.sheet.SheetDataSource} requires that the values returned
+	 *     by {@link #getGlobalStartIndex()} and {@link #getToIndex()} match the provided value. The
+	 *     value passed by {@link org.daisy.dotify.formatter.impl.sheet.SheetDataSource} is the
+	 *     number of pages that either the whole body, or the pre- or post-content of the current
+	 *     volume, already contains.
+	 * @param master
```
Comment (kalaspuffar) [inline src/org/daisy/dotify/formatter/impl/page/PageSequenceBuilder2.java]: I tried to build the project and when it builds javadoc it will fail due to a missing description for parameters. All of them need to be present if you started to document and they need a description to be allowed.

Comment (kalaspuffar) [reply src/org/daisy/dotify/formatter/impl/page/PageSequenceBuilder2.java]:
```
PageCounter.java:35: warning: no @return
        public int getDefaultPageOffset() {
                   ^
PageSequenceBuilder2.java:236: warning: no description for @param
         * @param transitionContent
           ^
PageSequenceBuilder2.java:237: warning: no description for @param
         * @param wasSplitInSequence
           ^
PageSequenceBuilder2.java:238: warning: no description for @param
         * @param isFirst
           ^
PageSequenceBuilder2.java:240: warning: no @return
        public PageImpl nextPage(int pageNumberOffset,
                        ^
PageSequenceBuilder2.java:700: warning: no @param for fromIndex
        public int getSizeLast(int fromIndex) {
                   ^
PageSequenceBuilder2.java:700: warning: no @return
        public int getSizeLast(int fromIndex) {
                   ^
PageSequenceBuilder2.java:712: warning: no @return
        public int size() {
                   ^
PageSequenceBuilder2.java:721: warning: no @return
        public int getGlobalStartIndex() {
                   ^
PageSequenceBuilder2.java:729: warning: no @return
        public int getToIndex() {
                   ^
PageSequenceBuilder2.java:53: error: invalid end tag: </br>
 *       <br></br>
             ^
PageSequenceBuilder2.java:112: warning: no description for @param
         * @param master
           ^
PageSequenceBuilder2.java:115: warning: no description for @param
         * @param seq
           ^
PageSequenceBuilder2.java:116: warning: no description for @param
         * @param context
           ^
PageSequenceBuilder2.java:117: warning: no description for @param
         * @param rcontext
           ^
PageSequenceBuilder2.java:118: warning: no description for @param
         * @param seqId
```

Comment (bertfrees) [reply src/org/daisy/dotify/formatter/impl/page/PageSequenceBuilder2.java]: Yes I got these warnings too but build didn't fail for me. I'll add the missing fields.

Comment (bertfrees) [reply src/org/daisy/dotify/formatter/impl/page/PageSequenceBuilder2.java]: By the way, just a side note, you should build with Java 11 to get the best results. There are a few things, such as linking to a package, that don't work with older versions.

File: src/org/daisy/dotify/formatter/impl/sheet/EvenSizeVolumeSplitter.java (hunk 1 above)
Comment (bertfrees) [reply src/org/daisy/dotify/formatter/impl/sheet/EvenSizeVolumeSplitter.java]: The way it works is not so intuitive I think, that can make it hard to understand. I've added a commit that explains things a bit more. It may also help to look at how EvenSizeVolumeSplitterCalculator works.

File: src/org/daisy/dotify/formatter/impl/search/VolumeKeepPriority.java [hunk 3]
```diff
@@ -89,6 +89,8 @@ public int orElse(int other) {
 	 * <p>Compares the specified values.</p>
 	 * <p>Note that, because a higher priority is represented by a lower value,
 	 * this function may appear to be incorrect in some contexts.</p>
+	 * <p>An absent priority is "higher" than the lowest priority ({@code 9}) is "higher" than the
+	 * highest priority ({@code 1}).</p>
```
Comment (PaulRambags) [inline src/org/daisy/dotify/formatter/impl/search/VolumeKeepPriority.java]: This sentence doesn't make sense.

**Comment by PaulRambags** [comment]
I already approved this PR but please resolve the conflicts and rephrase the sentence.

File: src/org/daisy/dotify/formatter/impl/search/VolumeKeepPriority.java (hunk 3 above)
Comment (bertfrees) [reply src/org/daisy/dotify/formatter/impl/search/VolumeKeepPriority.java]: Paul, maybe you can make a suggestion how it can be improved.

**Comment by bertfrees** [comment]
I resolved the conflicts (rebased onto master).

However, just to be clear, I think it is not "automatically" the job of the creator of the PR to resolve conflicts with things that have been added to master afterwards. Of course if the person who does the merges (Daniel) asks for help I'm happy to give it a try. In this case it was of course easy because the commit on master was also from me.

File: src/org/daisy/dotify/formatter/impl/search/VolumeKeepPriority.java (hunk 3 above)
Comment (PaulRambags) [reply src/org/daisy/dotify/formatter/impl/search/VolumeKeepPriority.java]: There are two problems with this sentence: 1. It is incorrect English, 2. it doesn't say what the absent priority is, except that it is higher than 9.
My suggestion to solve 1.: An absent priority is "higher" than the lowest priority (9) **and the lowest priority (9)** is "higher" than the highest priority (1).

Comment (bertfrees) [reply src/org/daisy/dotify/formatter/impl/search/VolumeKeepPriority.java]: Absent means it has no value. It's like an empty `Optional`.

**Comment by kalaspuffar** [comment]
Hi @bertfrees 

In this case, where we have a local repository where all members have access to change and update the PR we could do it. But in most cases, the merge request has been created in the git repository of the submitting party. For instance, I have a fork where I do my changes before submitting, and changes to that fork only I have access to update.

And much of the work done here is in the sbsdev fork and maintaining those pull request needs to be done by the submitting party or other interested developers at sbsdev. 

Best regards
Daniel

**Comment by bertfrees** [comment]
Well I guess it's a matter of agreeing on something. I just think it's not so self-evident. Why don't we bring it up in the next meeting...

**Comment by kalaspuffar** [comment]
Hi @bertfrees 

I've added the discussion point to the Meeting Agenda. 
Please verify the wording so we remember what needs discussing.

Best regards
Daniel

**Comment by bertfrees** [comment]
Perfect, thanks.

File: src/org/daisy/dotify/formatter/impl/search/VolumeKeepPriority.java (hunk 3 above)
Comment (kalaspuffar) [reply src/org/daisy/dotify/formatter/impl/search/VolumeKeepPriority.java]: Hi @bertfrees 

I agree with @PaulRambags that this sentence is a bit hard to parse.

But if I look at the code, it's more clear what the function does so let me try to understand the gist of it.

1. We are trying to compare how likely we should keep a volume is.
2. The lower the value, the higher the priority.
3. The values should range between 1-9, where 9 is the lowest priority.
4. If no value is present, we set the value to 10, which is even lower than the lowest priority (9).

As the function only has 1 line of code that you could understand, I think the importance of the comment is more to tell the reader what the ranges are and what we compare and why.

Best regards
Daniel

Comment (bertfrees) [reply src/org/daisy/dotify/formatter/impl/search/VolumeKeepPriority.java]: Hi Daniel. Your understanding of the class is correct. I think everything you said is in the Javadocs of the class and the various methods, you don't need to read any code to understand. Looking at it now I think we could say more clearly in the main Javadoc that a "high" priority means we try more to avoid splitting a volume within this area. How the range of values (1/9) are mapped to priorities (high/low priority) is mentioned several times throughout the class. The reason why I added another sentence along those same lines is to explain how the `compare` method behaves, because there are two possibilities and before it wasn't said which one it is. You can compare the values (9 > 1), or you can compare the priorities (high > low). In fact I think it would have been much more natural and intuitive if the latter would have been the case. It would have been much easier to write the documentation. However I wanted to focus on documentation only in this PR, I intentionally left code changes and refactorings for later. Note that the value 10 is only used internally in the compare method so that shouldn't be in the Javadoc IMO as to not confuse anyone that this would be the actual value.

By the way I notice now that the class does not implement `Comparator` or `Comparable`. But it should (preferably `Comparable`).

Comment (bertfrees) [reply src/org/daisy/dotify/formatter/impl/search/VolumeKeepPriority.java]: To me this is less clear than before. Sorry. My version makes it immediately clear to me how two `VolumeKeepPriority` objects are ordered. That information is also present in your version, but more difficult to extract IMO, its less unambiguous. Sure, my sentence could have been worded a bit better, but what I tried to achieve with it is lost now.

I think what I'll do tomorrow is change the behavior of the method the way I described in my previous comment. I think that should be done eventually, and then we don't waste so much time documenting this which is not easy. Agree? Or we can just drop the sentence altogether and I'll do it in another PR if you like.

Comment (kalaspuffar) [reply src/org/daisy/dotify/formatter/impl/search/VolumeKeepPriority.java]: Hi @bertfrees 

Sounds like a good idea. If you feel the patch is small and easy to review then add it. Otherwise drop the comment altogether and leave it to the next PR where we fix the functionality.

Best regards
Daniel

Comment (bertfrees) [reply src/org/daisy/dotify/formatter/impl/search/VolumeKeepPriority.java]: @kalaspuffar done.

**Comment by bertfrees** [comment]
@PaulRambags Don't worry, a new PR is coming soon.
